### ISSUE TITLE
This is for the pluggin ScaleUp, by Avidgamefan

### ISSUE THREAD

**Issue by MikeSavad** [easydiffusion/easydiffusion#1733]
I'd like 2 things:

1. More resolutions. The largest is taking forever. And its not that clear, I was happy at around 3200px, but right now its 3900 and up and it takes like 10-15min per image which is way too long.  So i'd like more choices right now it goes like 2200 to 3900,, i'd like it in the middle.

2. I'd like it to not age the people. When I enlarge a child with a small face, more often than not it makes the little girl old and wrinkled or it adds a mustache. 
![in_a_1920_s_doctor_exam_room__a_child_is_being_examined_by_a_kindly_ra_S2437887975_St30_G7 side](https://github.com/easydiffusion/easydiffusion/assets/88978360/62601600-0a7a-4ddb-a7af-7943d1b7e76d)

this is what i'm talking about, the girl on the left became old, the guy in the frame became young. It became one of those - what's different images, but I wanted it just bigger. I'd like to be able to adjust the weight of it to say  .1 or something maybe that will prevent it from happening.

**Comment by AvidGameFan** [comment]
If you want less-drastic changes as you scale up, there is an option to do that.  Click on the "Scale Up" label once, and it will change to "Scale Up (preserve)".  It uses a lower prompt-strength at that point, which will produce fewer changes.  Hopefully this, combined with the "sharpening" option will work better for you.

And, sometimes when you don't like the results, just scroll back to the previous image and try it again.  It uses a random seed, so each generation is unique.

As for the resolution choices, I've added another button to ScaleUp.  The first is what was there before, but the 2nd uses a 1.5x resize.  The 3rd remains the max.  Yes, the larger you go, the longer it will take.  I usually do 3 or 4 smaller resizes, and don't generally use the max size unless there's something particularly special.

In short, if you haven't downloaded a recent copy of ScaleUp, do so and give it another try.

**Comment by MikeSavad** [comment]
I don't think I ever saw that, but i'll check that again, I just go to the largest size and go from there. Though do wish there was one size under. Also the time estimate seems to be wrong. Sometimes it shows 15min but its more like 5.

Usually when it does its thing the amount shows 0.2

what I see right now are 3 sizes (which weird varies from model to model sometimes the max is 3900 and others are 4400) I see a 2x  and that grid box thing. 

For a while I downloaded it physically but it never updated. I have it set to download when updated and I saw it get updated but it doesn't look updated. Or at least I don't see the 1.5 or maybe i'm looking in the wrong spot. 

Also I doubt this is related to scaleup, but when I do a fix eyes -it removes all wrinkles, freckles etc. I don't get that.

**Comment by MikeSavad** [comment]
While i'm here does anyone know of other enlargers I can use besides the default one? I tried the 2X which took forever but seemed to have done a better job than the plugin (sorry plugin),  but it took a while. Is there a faster one?  I just like choices.

**Comment by MikeSavad** [comment]
Also there seems to be a bug?

When I hit 2X - it enlarges it, but when I hit download, it generates it all over again, even though I have the 2x size. I don't want it to do that.  Is that a bug or a feature and if its a feature can you disable that?

**Comment by AvidGameFan** [comment]
> I don't think I ever saw that, but i'll check that again, I just go to the largest size and go from there. Though do wish there was one size under.

There are 3 resolution buttons, now.  Don't go for the largest size first for best results. As I mentioned, I do multiple small ones before going for the large one, although it does take some time for each step.  The first button uses smaller steps, better for lower resolutions, or 1.2x for larger resolutions.  The 2nd button does a 1.5x resize.  The 3rd button just uses whatever the max. size is set to. 

You can change the max size by copying the plugin into the plugins folder and editing it.

> Also the time estimate seems to be wrong. Sometimes it shows 15min but its more like 5.

The time estimate is unrelated to ScaleUp.

> For a while I downloaded it physically but it never updated. I have it set to download when updated and I saw it get updated but it doesn't look updated. Or at least I don't see the 1.5 or maybe i'm looking in the wrong spot.

If you have a copy in your local plugin folder, it will not use the one in the plugin manager.

> While i'm here does anyone know of other enlargers I can use besides the default one? I tried the 2X which took forever but seemed to have done a better job than the plugin (sorry plugin), but it took a while. Is there a faster one? I just like choices.

There are 3 choices in the dropdown for upscalers.

My experience is the opposite -- the brute-force upscalers don't fully understand the context and don't generally do as good of a job.  But you have options.  Use what works for you.

> When I hit 2X - it enlarges it, but when I hit download, it generates it all over again, even though I have the 2x size. I don't want it to do that. Is that a bug or a feature and if its a feature can you disable that?

This is a feature, not a bug. If you do "2x", it runs the latant upscaler first, then run the normal img2img process -- two-in-one.  I don't think download has anything to do with it.

Hit the "Upscale" button for the built-in 2x or 4x upscale with no additional processing.

**Comment by MikeSavad** [comment]
Would it be possible to add another function button that does that multistep in one operation? If that gives better results (haven't tried it yet), click on button, multi step, largest size at the end. Is that something we can have?

I've heard of that step done in photoshop, to enlarge in small amounts and I found it works sometimes, but this is different.  Mostly I enlarge in batches and doing it one at a time would probably drive me insane. I'm guessing this would take a while though.

**Comment by AvidGameFan** [comment]
> Would it be possible to add another function button that does that multistep in one operation? If that gives better results (haven't tried it yet), click on button, multi step, largest size at the end. Is that something we can have?

It's possible to run multiple generates with one click,, but ED still likes to run each generation separately, so it would be confusing to see it automatically kick off subsequent runs. And it'll save all of the intermediate images. Users would have to be able to deal with that, or I'd have to come up with a way to display some sort of message -- I dunno. Also, I'm not sure how to easily cancel the processing, but I suppose I could figure that out.  But yeah, it can be done.

The thing is, sometimes when you do these incremental/intermediate upscales, it can introduce an odd artifact (such as changing the hands in a bad way, adding a finger, or whatever).  So, normally, it's easy enough to just re-run and verify that you're happy with the results before continuing to scale-up.  If you do it in one step, you may have to start over, after having watched it generate a few, so being able to cancel seems important to me.

I increased the max size.  Personally, I rarely generate the max size, as most of the time I find 6mp to be good enough for me.  It just takes a lot longer the larger you make it.  So, there probably should be an option to multi-ScaleUp part way or all the way, or some other way to limit based on preference.

**Comment by MikeSavad** [comment]
I tried do it in smaller runs and while it looked better, after 20min I gave up bumping it in small amounts. It was a sort of hair pulling experience. 

I found the older version that stopped at 3200px was enough for me, but that option is gone and replaced by 4400 which looks more like it was enlarged in general. Mostly I want the 3200 size (I have mine at 1600 on a run). Or to have an option of just different stops -  2000 - 3200 - then the largest.

**Comment by AvidGameFan** [comment]
> I tried do it in smaller runs and while it looked better, after 20min I gave up bumping it in small amounts. It was a sort of hair pulling experience.

I often work on 2 to 4 at the same time, queue up some runs, then go do something else or work on the prompt.  With the Turbo model, it takes like 30 seconds to a minute to make a run at the lower resolutions, so there's not much wait at first.  It takes increasingly longer to generate at higher resolutions.

> I found the older version that stopped at 3200px was enough for me, but that option is gone and replaced by 4400 which looks more like it was enlarged in general. Mostly I want the 3200 size (I have mine at 1600 on a run). Or to have an option of just different stops - 2000 - 3200 - then the largest.

I upped the Max size, per request.  This is the thing -- everyone has their own target maximum.  Some people have a faster card and don't mind going larger.  Myself, I usually just stop when I get to 6mp.  By using an intermediate scale-up, you can stop wherever you want.  You're not going to get best results with just one scale-up all the way to max anyway.

If you copy the plugin file (scaleup.plugin.js) to your local plugin folder, you can edit the max size. Use a plain text editor (like notepad) and it should be easy to see.  And actually, you can edit the scaling factors if you want larger jumps for the intermediate buttons.  (Currently set to 1.25 and 1.5.  1.5 is a big enough jump, tho, IMHO.)

I've been thinking for a while about automating the multi-generate process, but it's just tricky enough to not be straightforward.

**Comment by AvidGameFan** [comment]
OK, I went ahead and added the multi-scaleup feature.  It's in a basic form right now -- I have it hard-coded to 3 scaleups.  After the 3, the result should be 5.5 times the resolution.  So, if you start with a 1mp image, you'll end up around 5.5mp.  This could be increased, but is a good start to see how it goes.

I've been thinking for a while how it might work, and it ended up working very close to how I imagined it.  I had some hurdles to jump, but with the flexible architecture of ED and the modular approach I used in the plugin, it was easier than I expected.

**Comment by MikeSavad** [comment]
Cool thanks, i'll have to give it a try later on.   Right now, or pre this update, 2x worked pretty well and was pretty fast. The largest size took pretty long. And the results were mixed, usually it had the look of an enlarged image but with added details. So i'll try this one. Thanks

----Mike Savad

**Comment by MikeSavad** [comment]
Judging by the time the program says (which is suspect), but the total time was like 3 min, then 6, then 15? I think it was about half that. The multistep takes a while but I don't need to be there which is good. And the results do look pretty good. I'm comparing it to the highest jump next.

The results took longer on the multi but the results were a lot cleaner. Multi gave me a final image of around 3700px from 1600px. The big jump was 3900 (on some models its 4400?).

The results on the new one were a lot cleaner, but resulted in  a lot more images on my screen that I had to sort through. 
![multistep](https://github.com/easydiffusion/easydiffusion/assets/88978360/4a342cd6-6a40-41bc-8539-78dbbbfe2b5e)

the fine lines swirls in the sky look cleaner and I don't get those enlarging compression outlines on the hair with the new version. It also looks brighter and sharper. On the negative? It added flowers were there weren't any before, removed items, the moon is full and not part. The clouds are different. Its not a bad thing I wouldn't have noticed it if I wasn't staring but it did change it a bit from original. It also added that face in the moon. The full leap is more true to the original but it does lack a crispness.

**Comment by AvidGameFan** [comment]
> Judging by the time the program says (which is suspect), but the total time was like 3 min, then 6, then 15? I think it was about half that. 
I think the time reported may be cumulative?  Anyway, if you want to run faster, try a Turbo model with 7 steps.
> The multistep takes a while but I don't need to be there which is good. And the results do look pretty good. 
> ...The results took longer on the multi but the results were a lot cleaner.

Yes, that's the goal!  Superior results.

> ... but resulted in a lot more images on my screen that I had to sort through.

Yes.  But now you have the option to revisit intermediate versions if necessary.  Like, if you didn't like how it changed the moon or added flowers, you could go back to a version before that any try scaling up (where it uses different seeds each time) to see if you get something else.

So, I see the intermediate files as an advantage.  However, as a practical matter, I'm working with the existing design of ED, and I think it'd be a lot more difficult to generate 3 behind the scenes without any intermediate files.

To help mitigate the large number of files that I generate (whether or not using multi), I created a "Favorites" plugin and file-copy utility.  This is going to save me a lot of time in selecting only the ones I think are worthy.  (If you have the option for it turned on, ED will save every generated image.  Another option is to turn the automatic save off, and you won't have any image saved unless you explicitly select to download it.

> the fine lines swirls in the sky look cleaner and I don't get those enlarging compression outlines on the hair with the new version. It also looks brighter and sharper. On the negative? It added flowers were there weren't any before, removed items, the moon is full and not part. The clouds are different. Its not a bad thing I wouldn't have noticed it if I wasn't staring but it did change it a bit from original. It also added that face in the moon. The full leap is more true to the original but it does lack a crispness.

Can flowers actually be a negative?  😅 Overall, it looks more coherent and if asked to choose between the two, I'd take the multi, no question.  But if you try it again, it'll do something different.  You could also put "flowers" in the negative prompt and do it over.  Sometimes you just have to intervene and do manual steps, if the AI is stubborn.

What's nice about the multi process is:

1. Gaining resolution
2. Gaining detail
3. Queueing up a couple of runs, and doing something else for a few minutes

You really don't want #1 without #2.  Or if you do, just use the 2x or 4x Upscalers.

There is an option to turn off the sharpening (see the panel in the left column, at the bottom), if you don't feel that it is needed.  If you don't run the multi, but just scale up incrementally by hand, you can choose at each step whether or not to add the sharpening.

**Comment by MikeSavad** [comment]
I wonder how good the results would be if it was only 2 steps. I did a batch of like 5 things and it took at least an hour. 

I forgot about that preserve thing. In my early days I clicked on stuff. Now I just poke things. 

I'm also trying to figure out why sometimes my total size is 4400px and other models is 3900px or so. 

That sharpening, where is that located? I probably have too many plugin icons on my screen to see it. (Really need to try them out some day I guess).  Is the sharpening on the final image?

**Comment by MikeSavad** [comment]
One other thing, right now if I do a batch of different images, it goes through all images making all sizes. And its real confusing figuring out which ones are done. Could it be done as a single batch.

All cats
all cars etc so its a stack of images that are the same so I know if there are 3 in a row, the top one is done. 


----Mike Savad

**Comment by AvidGameFan** [comment]
> I wonder how good the results would be if it was only 2 steps. I did a batch of like 5 things and it took at least an hour.

Well, you can manually click on one of the scaleup buttons on each image, wander away from your machine, come back in a few and click on the next button all over again on all of the pending images.  Or you can just try it out and see. 

Even 2 scale-ups are better than none.  4 is even better than 3, I think.

> I'm also trying to figure out why sometimes my total size is 4400px and other models is 3900px or so.

It just scales-up based on your starting resolution and aspect ratio.  So, if you're wider, or start larger, it'll end up larger.

> That sharpening, where is that located? I probably have too many plugin icons on my screen to see it. (Really need to try them out some day I guess). Is the sharpening on the final image?

Go all the way left on your screen, off the image, to the settings panels.  Scroll down until you see the ScaleUp settings.  Click on it to open, if it's not already open.

> One other thing, right now if I do a batch of different images, it goes through all images making all sizes. And its real confusing figuring out which ones are done. Could it be done as a single batch.

As I said, I think it'd be a lot more difficult to hide the intermediate runs.  Not sure how to do it in a plugin, without modifying ED.  There could be a way, I just don't know it.

> All cats
> all cars etc so its a stack of images that are the same so I know if there are 3 in a row, the top one is done.

If your images look different enough, you can see which was first (smallest resolution, lower down), and which was last (largest resolution, further up).  If you are in the file system, you can sort by date and also verify looking at the file sizes on disk -- smallest size is usually the first one, etc.  But this problem exists even before the multi-generate, at least for me.  This just further automates what I do anyway -- multiple scale-ups, with detours to inpaint, do-over, etc.

**Comment by MikeSavad** [comment]
Its weird I tried it out today, but I got some error pop up on the screen but it went by too fast to read. After which I had 1 getting larger with a picture and 2 that were blank. I wasn't sure if that was right so I erased those. I went back to try it again, clicked the 123 button and nothing happened. Like it was a one time deal. Maybe its just my memory so it could be that.

**Comment by AvidGameFan** [comment]
There could be bugs, or, more likely, some weird combination of things that I didn't expect that can cause problems.  But it's been working reliably for me.  ED puts a lot of log information in the command window that runs in the background, so check that for errors too.

**Comment by MikeSavad** [comment]
What does happen now and then is when I go to copy the text to save with my file, now and then it crashes the whole window. I can't tell if that's the pluggin, the program or vivaldi the browser. But I only copy it after its enlarged, so i'm thinking its the program.  Do you know where that log file is? By any chance?

**Comment by AvidGameFan** [comment]
If you have errors while rendering, you should find more details in the Command Prompt that is running the server.  It looks something like this:

![image](https://github.com/easydiffusion/easydiffusion/assets/60022188/4e2b71ff-f111-4c9d-b5d9-b5d314870c83)

This is not part of Vivaldi.  Not sure what you're doing there.

**Comment by MikeSavad** [comment]
I'll have to look, all that stuff just wizzes by, it still keeps making stuff sometimes, even without the browser.

As I was playing with the enlarging I found 2x works but it eats all the memory, like it enlarges then it processes and then goes back to the other images and eventually crashes. I wonder if the 123 step should be  just 2 steps to double the size. Or if the double size could just process the image without doing it twice.

**Comment by AvidGameFan** [comment]
The new 3-step icon uses 3 smaller steps, as you get more detail the more rounds you use.  If you need it larger, you can run the simple upscale when it's done.  Or run it through more scale-up steps.  Or hit the "max" resolution and get more than 2x size.  

All of the options are compromises.  The trick is finding which compromises give you the best results for what you're trying to do.  The 2x is a compromise meant for when you want to do a 2x scaleup with the fewest steps possible.  It won't eat all of the memory if you start from lower resolutions.  It really depends on how much VRAM you have.  I optimized it for 8GB VRAM, as that's what's on my machine.  Note that for SDXL you also want to run it with the "low" GPU setting.

**Comment by AvidGameFan** [comment]
Also, yeah -- if your browser crashes for whatever reason, the server can still finish whatever image it's generating.  If you have auto-save turned out, it'll still get written to your storage folder.

I can elaborate on the reasoning behind some of my choices around the 2x.  It would be easy to another resolution button for 2x (and even easier if you want to change the 1.5x button to 2x), but the bigger the jump, when using img2img, the less detail you get.  That's why the existing buttons are 1.25x and 1.5x.  2x just won't give you as much detail, so I leveraged the use of the latent upscaler to do the initial 2x, which enhances the image a bit, then img2img on top of that to add detail, which works better than just doing 2x straight.

But if you want to give it a try, copy the plugin code into your local plugins folder, and edit this line:

var scalingIncrease2=1.5;

You should be able to put any factor you want in there.  2.2x if you want!
